samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -0,0 +1,221 @@
+  }
+
+  @Override
+  public void onProcessorChange(List<String> processorIds) {
+    log.info("ZkJobCoordinator::onProcessorChange - Processors changed! List: " + Arrays.toString(processorIds.toArray()));
+    generateNewJobModel();
+  }
+
+  @Override
+  public void onNewJobModelAvailable(final String version) {
+    newJobModelVersion = version;
+    log.info("pid=" + processorId + "new JobModel available");
+    // stop current work
+    containerController.stopContainer();
+    log.info("pid=" + processorId + "new JobModel available.Container stopped.");
+    // get the new job model
+    newJobModel = zkUtils.getJobModel(version);
+    log.info("pid=" + processorId + "new JobModel available. ver=" + version + "; jm = " + newJobModel);
+
+    String currentPath = zkUtils.getEphemeralPath();
+    String zkProcessorId = keyBuilder.parseIdFromPath(currentPath);
+
+    // update ZK and wait for all the processors to get this new version
+    barrier.waitForBarrier(version, String.valueOf(zkProcessorId), new Runnable() {
+      @Override
+      public void run() {
+        onNewJobModelConfirmed(version);
+      }
+    });
+  }
+
+  @Override
+  public void onNewJobModelConfirmed(String version) {
+    log.info("pid=" + processorId + "new version " + version + " of the job model got confirmed");
+    // get the new Model
+    JobModel jobModel = getJobModel();
+    log.info("pid=" + processorId + "got the new job model in JobModelConfirmed =" + jobModel);
+
+    // start the container with the new model
+    containerController.startContainer(jobModel.getContainers().get(processorId), jobModel.getConfig(),
+        jobModel.maxChangeLogStreamPartitions);
+  }
+
+  /**
+   * Generate new JobModel when becoming a leader or the list of processor changed.
+   */
+  private void generateNewJobModel() {
+    // get the current list of processors
+    List<String> currentProcessors = zkUtils.getSortedActiveProcessors();
+
+    // get the current version
+    String currentJMVersion  = zkUtils.getJobModelVersion();
+    String nextJMVersion;
+    if (currentJMVersion == null) {
+      log.info("pid=" + processorId + "generating first version of the model");
+      nextJMVersion = "1";
+    } else {
+      nextJMVersion = Integer.toString(Integer.valueOf(currentJMVersion) + 1);
+    }
+    log.info("pid=" + processorId + "generating new model. Version = " + nextJMVersion);
+
+    StringBuilder sb = new StringBuilder();
+    List<Integer> containerIds = new ArrayList<>();
+    for (String processor : currentProcessors) {
+      String zkProcessorId = keyBuilder.parseIdFromPath(processor);
+      sb.append(zkProcessorId).append(",");
+      containerIds.add(Integer.valueOf(zkProcessorId));
+    }
+    log.info("generate new job model: processorsIds: " + sb.toString());
+
+    jobModelManager = JobModelManager$.MODULE$.getJobCoordinator(this.config, null, null, streamMetadataCache, null,
```

Comment:
Based on this statement, it seems like we are not supporting localitymanager and changelog manager. So, are we explicitly saying we are not going to support stateful jobs in this version? 
Thanks to @santhoshvenkat1988 for pointing it out.